File: leetcode/array/33.SearchInRotatedSortedArray/33SearchInRotatedSortedArray.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        if not len(nums):
            return -1
        left, right = 0, len(nums)-1
        while left+1 < right:
            mid = left + (right - left) // 2
            if nums[mid] == target:
                return mid
            if nums[left] < nums[mid]:
                if nums[left] <= target <= nums[mid]:
                    right = mid
                else:
                    left = mid
            elif nums[mid] < nums[right]:
                if nums[mid] <= target <= nums[right]:
                    left = mid
                else:
                    right = mid
        #end_while
        if nums[left] == target:
            return left
        if nums[right] == target:
            return right
        return -1

File: leetcode/array/33.SearchInRotatedSortedArray/33SearchInRotatedSortedArray.js
var search = function(nums, target) {
    if (nums.length == 0) return -1;
    if (nums.length == 1) return nums[0] == target ? 0 : -1;
    var l = 0;
    var r = nums.length - 1;
    while (l < r) {
        if (nums[l] == target) {
            return l;
        } else if (nums[r] == target) {
            return r;
        }

        if (nums[l] > target) {
            while (nums[r] == nums[r - 1]) {
                r--;
            }
            if (nums[r] < target) {
                return -1;
            } else if (nums[r] > target) {
                r--;
            } else if (nums[r] == target) {
                return r;
            }
        } else if (nums[l] < target) {
            // 跳过数组一样的
            while (nums[l] == nums[l + 1]) {
                l++;
            }

            if (nums[l] < nums[l + 1]) {
                l++;
            } else {
                return -1;
            }
        } else if (nums[l] < nums[r]) {
            return -1;
        }
    }
    return -1;
};
